### PR TITLE
Clean up after a failure in a run callback

### PR DESCRIPTION
**Pull request by matthewd**

### CODE REVIEW

File: activerecord/lib/active_record/query_cache.rb [hunk 1]
```diff
@@ -23,23 +23,27 @@ def uncached(&block)
       end
     end
 
-    def self.install_executor_hooks(executor = ActiveSupport::Executor)
-      executor.to_run do
-        connection    = ActiveRecord::Base.connection
-        enabled       = connection.query_cache_enabled
-        connection_id = ActiveRecord::Base.connection_id
-        connection.enable_query_cache!
+    def self.run
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/query_cache.rb]: One question about this. is `run` executede when the application boots?

Comment (matthewd) [inline activerecord/lib/active_record/query_cache.rb]: No; the executor calls `run`/`complete` strictly in pairs around "user" code.

File: activesupport/test/executor_test.rb [hunk 2]
```diff
@@ -35,6 +38,20 @@ def test_separated_calls_invoke_callbacks
...
+    executor.to_run { raise DummyError }
+    executor.to_run { called << :run_2 }
+    executor.to_complete { called << :complete }
+
+    assert_raises(DummyError) do
+      executor.wrap { called << :body }
+    end
+
+    assert_equal [:run_1, :complete], called
+  end
```
Comment (jeremy) [inline activesupport/test/executor_test.rb]: 🤘

File: activesupport/lib/active_support/execution_wrapper.rb [hunk 3]
```diff
@@ -19,6 +19,32 @@ def self.to_complete(*args, &block)
...
+      to_run(*run_args) do
+        hook_state[hook] = hook.run
+      end
+      to_complete(*complete_args) do
+        if hook_state.key?(hook)
+          hook.complete hook_state[hook]
+        end
+      end
+    end
+
```
Comment (jeremy) [inline activesupport/lib/active_support/execution_wrapper.rb]: Nice API. The `outer:` arg is mostly self-explanatory but suggests some commentary on run/complete order.

**Comment by rafaelfranca** [comment]
:+1:

**Pull request merged**